From 89e007dcf6891140f0fe7bfa58eb27b8da4fcb7d Mon Sep 17 00:00:00 2001
From: Chau Ly <chaul@amperecomputing.com>
Date: Thu, 26 Jun 2025 03:29:52 +0000
Subject: [PATCH] platform-mc: Support PLDM File Transfer commands

This commit supports functions for PLDM File Transfer command, which
enlists PLDM File Transfer command encode/decode APIs in libpldm. They
includes: DfOpen, DfClose, MultipartReceive, DfHeartbeat,
NegotiateTransferParameters commands.

Upstream-Status: Submitted [https://gerrit.openbmc.org/c/openbmc/pldm/+/82471]

Change-Id: Idc747c43fd35abf34e54b9632782f0ee0d03a980
Signed-off-by: Chau Ly <chaul@amperecomputing.com>
(cherry picked from commit 1957c07eea1142fd5ff707a311e5eacddbc3c5a5)
---
 common/types.hpp          |   1 +
 meson.build               |   1 +
 platform-mc/file_cmds.cpp | 277 ++++++++++++++++++++++++++++++++++++++
 platform-mc/file_cmds.hpp |  96 +++++++++++++
 4 files changed, 375 insertions(+)
 create mode 100644 platform-mc/file_cmds.cpp
 create mode 100644 platform-mc/file_cmds.hpp

diff --git a/common/types.hpp b/common/types.hpp
index 70fe09d..1e387ae 100644
--- a/common/types.hpp
+++ b/common/types.hpp
@@ -288,6 +288,7 @@ using Callback = std::function<void()>;
 namespace file_transfer
 {
 using FileID = uint16_t;
+using CRC32Type = uint32_t;
 
 enum class FileClassification : uint8_t
 {
diff --git a/meson.build b/meson.build
index 9bcd2f6..af9e2a9 100644
--- a/meson.build
+++ b/meson.build
@@ -324,6 +324,7 @@ executable(
     'platform-mc/sensor_manager.cpp',
     'platform-mc/numeric_sensor.cpp',
     'platform-mc/event_manager.cpp',
+    'platform-mc/file_cmds.cpp',
     'platform-mc/dbus_to_terminus_effecters.cpp',
     oem_files,
     'requester/mctp_endpoint_discovery.cpp',
diff --git a/platform-mc/file_cmds.cpp b/platform-mc/file_cmds.cpp
new file mode 100644
index 0000000..527902f
--- /dev/null
+++ b/platform-mc/file_cmds.cpp
@@ -0,0 +1,277 @@
+#include "file_cmds.hpp"
+
+PHOSPHOR_LOG2_USING;
+
+namespace pldm
+{
+namespace platform_mc
+{
+namespace file_cmds
+{
+
+exec::task<int> dfOpen(TerminusManager& terminusManager, pldm_tid_t tid,
+                       const struct pldm_file_df_open_req* reqPtr,
+                       struct pldm_file_df_open_resp* respPtr)
+{
+    size_t payload_length = PLDM_DF_OPEN_REQ_BYTES;
+    Request request(sizeof(pldm_msg_hdr) + payload_length);
+    auto requestMsg = new (request.data()) pldm_msg;
+
+    auto rc =
+        encode_pldm_file_df_open_req(0, reqPtr, requestMsg, &payload_length);
+
+    if (rc)
+    {
+        lg2::error(
+            "Failed to encode DfOpen request for terminus ID {TID}, FileIdentifier {ID}, error {RC} ",
+            "TID", tid, "ID", reqPtr->file_identifier, "RC", rc);
+        co_return PLDM_ERROR;
+    }
+
+    const pldm_msg* responseMsg = nullptr;
+    size_t responseLen = 0;
+    rc = co_await terminusManager.sendRecvPldmMsg(tid, request, &responseMsg,
+                                                  &responseLen);
+    if (rc)
+    {
+        lg2::error(
+            "Failed to send DfOpen request for terminus ID {TID}, FileIdentifier {ID}, error {RC} ",
+            "TID", tid, "ID", reqPtr->file_identifier, "RC", rc);
+        co_return rc;
+    }
+
+    rc = decode_pldm_file_df_open_resp(responseMsg, responseLen, respPtr);
+
+    if (rc)
+    {
+        lg2::error(
+            "Failed to decode DfOpen response for terminus ID {TID}, FileIdentifier {ID}, error {RC} ",
+            "TID", tid, "ID", reqPtr->file_identifier, "RC", rc);
+        co_return PLDM_ERROR;
+    }
+
+    if (respPtr->completion_code)
+    {
+        lg2::error(
+            "Failed to send DfOpen request for terminus ID {TID}, FileIdentifier {ID}, completion code {CC}",
+            "TID", tid, "ID", reqPtr->file_identifier, "CC",
+            respPtr->completion_code);
+        co_return respPtr->completion_code;
+    }
+
+    co_return respPtr->completion_code;
+}
+
+exec::task<int> dfClose(TerminusManager& terminusManager, pldm_tid_t tid,
+                        const struct pldm_file_df_close_req* reqPtr)
+{
+    size_t payload_length = PLDM_DF_CLOSE_REQ_BYTES;
+    Request request(sizeof(pldm_msg_hdr) + payload_length);
+    auto requestMsg = new (request.data()) pldm_msg;
+
+    auto rc =
+        encode_pldm_file_df_close_req(0, reqPtr, requestMsg, &payload_length);
+    if (rc)
+    {
+        lg2::error(
+            "Failed to encode DfClose request for terminus ID {TID}, file descriptor {FD}, error {RC} ",
+            "FD", reqPtr->file_descriptor, "RC", rc);
+        co_return PLDM_ERROR;
+    }
+
+    const pldm_msg* responseMsg = nullptr;
+    size_t responseLen = 0;
+    rc = co_await terminusManager.sendRecvPldmMsg(tid, request, &responseMsg,
+                                                  &responseLen);
+    if (rc)
+    {
+        lg2::error(
+            "Failed to send DfClose request for terminus ID {TID}, file descriptor {FD}, error {RC} ",
+            "TID", tid, "FD", reqPtr->file_descriptor, "RC", rc);
+        co_return rc;
+    }
+
+    struct pldm_file_df_close_resp resp_data = {};
+    rc = decode_pldm_file_df_close_resp(responseMsg, responseLen, &resp_data);
+    if (rc)
+    {
+        lg2::error(
+            "Failed to decode DfClose response for terminus ID {TID}, file descriptor {FD}, error {RC} ",
+            "TID", tid, "FD", reqPtr->file_descriptor, "RC", rc);
+        co_return PLDM_ERROR;
+    }
+
+    if (resp_data.completion_code)
+    {
+        lg2::error(
+            "Failed to send DfClose request for terminus ID {TID}, file descriptor {FD}, completion code {CC}",
+            "TID", tid, "FD", reqPtr->file_descriptor, "CC",
+            resp_data.completion_code);
+        co_return resp_data.completion_code;
+    }
+
+    co_return resp_data.completion_code;
+}
+
+exec::task<int> dfHeartbeat(TerminusManager& terminusManager, pldm_tid_t tid,
+                            const struct pldm_file_df_heartbeat_req* reqPtr,
+                            struct pldm_file_df_heartbeat_resp* respPtr)
+{
+    size_t payload_length = PLDM_DF_HEARTBEAT_REQ_BYTES;
+    Request request(sizeof(pldm_msg_hdr) + payload_length);
+    auto requestMsg = new (request.data()) pldm_msg;
+
+    auto rc = encode_pldm_file_df_heartbeat_req(0, reqPtr, requestMsg,
+                                                &payload_length);
+    if (rc)
+    {
+        lg2::error(
+            "Failed to encode DfHeartbeat request for terminus ID {TID}, file descriptor {FD}, error {RC} ",
+            "TID", tid, "FD", reqPtr->file_descriptor, "RC", rc);
+        co_return PLDM_ERROR;
+    }
+
+    const pldm_msg* responseMsg = nullptr;
+    size_t responseLen = 0;
+    rc = co_await terminusManager.sendRecvPldmMsg(tid, request, &responseMsg,
+                                                  &responseLen);
+    if (rc)
+    {
+        lg2::error(
+            "Failed to send DfHeartbeat request for terminus ID {TID}, file descriptor {FD}, error {RC} ",
+            "TID", tid, "FD", reqPtr->file_descriptor, "RC", rc);
+        co_return rc;
+    }
+
+    rc = decode_pldm_file_df_heartbeat_resp(responseMsg, responseLen, respPtr);
+    if (rc)
+    {
+        lg2::error(
+            "Failed to decode DfHeartbeat response for terminus ID {TID}, file descriptor {FD}, error {RC} ",
+            "TID", tid, "FD", reqPtr->file_descriptor, "RC", rc);
+        co_return PLDM_ERROR;
+    }
+
+    if (respPtr->completion_code)
+    {
+        lg2::error(
+            "Failed to send DfHeartbeat request for terminus ID {TID}, file descriptor {FD}, completion code {CC}",
+            "TID", tid, "FD", reqPtr->file_descriptor, "CC",
+            respPtr->completion_code);
+        co_return respPtr->completion_code;
+    }
+
+    co_return respPtr->completion_code;
+}
+
+exec::task<int> negotiateTransferParameters(
+    TerminusManager& terminusManager, pldm_tid_t tid,
+    const struct pldm_base_negotiate_transfer_params_req* reqPtr,
+    struct pldm_base_negotiate_transfer_params_resp* respPtr)
+{
+    size_t payload_length = PLDM_BASE_NEGOTIATE_TRANSFER_PARAMETERS_REQ_BYTES;
+    Request request(sizeof(pldm_msg_hdr) + payload_length);
+    auto requestMsg = new (request.data()) pldm_msg;
+
+    auto rc = encode_pldm_base_negotiate_transfer_params_req(
+        0, reqPtr, requestMsg, &payload_length);
+    if (rc)
+    {
+        lg2::error(
+            "Failed to encode NegotiateTransferParameters request for terminus ID {TID}, error {RC} ",
+            "TID", tid, "RC", rc);
+        co_return PLDM_ERROR;
+    }
+
+    const pldm_msg* responseMsg = nullptr;
+    size_t responseLen = 0;
+    rc = co_await terminusManager.sendRecvPldmMsg(tid, request, &responseMsg,
+                                                  &responseLen);
+    if (rc)
+    {
+        lg2::error(
+            "Failed to send NegotiateTransferParameters request for terminus ID {TID}, error {RC} ",
+            "TID", tid, "RC", rc);
+        co_return rc;
+    }
+
+    rc = decode_pldm_base_negotiate_transfer_params_resp(responseMsg,
+                                                         responseLen, respPtr);
+    if (rc)
+    {
+        lg2::error(
+            "Failed to decode NegotiateTransferParameters response for terminus ID {TID}, error {RC} ",
+            "TID", tid, "RC", rc);
+        co_return PLDM_ERROR;
+    }
+
+    if (respPtr->completion_code)
+    {
+        lg2::error(
+            "Failed to send NegotiateTransferParameters request for terminus ID {TID}, completion code {CC}",
+            "TID", tid, "CC", respPtr->completion_code);
+        co_return respPtr->completion_code;
+    }
+
+    co_return respPtr->completion_code;
+}
+
+exec::task<int> sendMultipartReceive(
+    TerminusManager& terminusManager, pldm_tid_t tid,
+    const struct pldm_base_multipart_receive_req* reqPtr,
+    struct pldm_base_multipart_receive_resp* respPtr,
+    CRC32Type& dataIntegrityChecksum)
+{
+    size_t payload_length = PLDM_MULTIPART_RECEIVE_REQ_BYTES;
+    Request request(sizeof(pldm_msg_hdr) + payload_length);
+    auto requestMsg = new (request.data()) pldm_msg;
+
+    auto rc = encode_pldm_base_multipart_receive_req(0, reqPtr, requestMsg,
+                                                     &payload_length);
+    if (rc)
+    {
+        lg2::error(
+            "Failed to encode MultipartReceive request for terminus ID {TID}, file descriptor {FD}, error {RC} ",
+            "TID", tid, "FD", reqPtr->transfer_ctx, "RC", rc);
+        co_return PLDM_ERROR;
+    }
+
+    const pldm_msg* responseMsg = nullptr;
+    size_t responseLen = 0;
+    rc = co_await terminusManager.sendRecvPldmMsg(tid, request, &responseMsg,
+                                                  &responseLen);
+    if (rc)
+    {
+        lg2::error(
+            "Failed to send MultipartReceive request for terminus ID {TID}, file descriptor {FD}, error {RC} ",
+            "TID", tid, "FD", reqPtr->transfer_ctx, "RC", rc);
+        co_return rc;
+    }
+
+    rc = decode_pldm_base_multipart_receive_resp(
+        responseMsg, responseLen, respPtr, &dataIntegrityChecksum);
+    if (rc)
+    {
+        lg2::error(
+            "Failed to decode MultipartReceive response for terminus ID {TID}, file descriptor {FD}, error {RC} ",
+            "TID", tid, "FD", reqPtr->transfer_ctx, "RC", rc);
+        co_return PLDM_ERROR;
+    }
+
+    if (respPtr->completion_code)
+    {
+        lg2::error(
+            "Failed to send MultipartReceive request for terminus ID {TID}, file descriptor {FD}, completion code {CC}",
+            "TID", tid, "FD", reqPtr->transfer_ctx, "CC",
+            respPtr->completion_code);
+        co_return respPtr->completion_code;
+    }
+
+    co_return respPtr->completion_code;
+}
+
+} // namespace file_cmds
+
+} // namespace platform_mc
+
+} // namespace pldm
diff --git a/platform-mc/file_cmds.hpp b/platform-mc/file_cmds.hpp
new file mode 100644
index 0000000..c9d5a07
--- /dev/null
+++ b/platform-mc/file_cmds.hpp
@@ -0,0 +1,96 @@
+#pragma once
+
+#include "common/transport.hpp"
+#include "common/utils.hpp"
+#include "terminus.hpp"
+#include "terminus_manager.hpp"
+
+#include <libpldm/base.h>
+#include <libpldm/file.h>
+
+#include <phosphor-logging/lg2.hpp>
+
+#include <functional>
+
+namespace pldm
+{
+
+namespace platform_mc
+{
+
+namespace file_cmds
+{
+using namespace pldm::file_transfer;
+
+/** @brief Send DfOpen command to file host
+ *
+ *  @param[in] terminusManager - TerminusManager object
+ *  @param[in] tid - Destination TID
+ *  @param[in] reqPtr - pointer to the request struct
+ *  @param[out] respPtr - pointer to the response struct
+ *
+ *  @return PLDM completion code
+ */
+exec::task<int> dfOpen(TerminusManager& terminusManager, pldm_tid_t tid,
+                       const struct pldm_file_df_open_req* reqPtr,
+                       struct pldm_file_df_open_resp* respPtr);
+
+/** @brief Send DfClose command to file host
+ *
+ *  @param[in] terminusManager - TerminusManager object
+ *  @param[in] tid - Destination TID
+ *  @param[in] reqPtr - pointer to the request struct
+ *
+ *  @return PLDM completion code
+ */
+exec::task<int> dfClose(TerminusManager& terminusManager, pldm_tid_t tid,
+                        const struct pldm_file_df_close_req* reqPtr);
+
+/** @brief Send DfHeartbeat command to file host
+ *
+ *  @param[in] terminusManager - TerminusManager object
+ *  @param[in] tid - Destination TID
+ *  @param[in] reqPtr - pointer to the request struct
+ *  @param[out] respPtr - pointer to the response struct
+ *
+ *  @return PLDM completion code
+ */
+exec::task<int> dfHeartbeat(TerminusManager& terminusManager, pldm_tid_t tid,
+                            const struct pldm_file_df_heartbeat_req* reqPtr,
+                            struct pldm_file_df_heartbeat_resp* respPtr);
+
+/** @brief Send NegotiateTransferParameters command to file host
+ *
+ *  @param[in] terminusManager - TerminusManager object
+ *  @param[in] tid - Destination TID
+ *  @param[in] reqPtr - pointer to the request struct
+ *  @param[out] respPtr - pointer to the response struct
+ *
+ *  @return PLDM completion code
+ */
+exec::task<int> negotiateTransferParameters(
+    TerminusManager& terminusManager, pldm_tid_t tid,
+    const struct pldm_base_negotiate_transfer_params_req* reqPtr,
+    struct pldm_base_negotiate_transfer_params_resp* respPtr);
+
+/** @brief Send MultipartReceive command to file host
+ *
+ *  @param[in] terminusManager - TerminusManager object
+ *  @param[in] tid - Destination TID
+ *  @param[in] reqPtr - pointer to the request struct
+ *  @param[out] respPtr - pointer to the response struct
+ *  @param[out] dataIntegrityChecksum - DataIntegrityChecksum
+ *
+ *  @return PLDM completion code
+ */
+exec::task<int> sendMultipartReceive(
+    TerminusManager& terminusManager, pldm_tid_t tid,
+    const struct pldm_base_multipart_receive_req* reqPtr,
+    struct pldm_base_multipart_receive_resp* respPtr,
+    CRC32Type& dataIntegrityChecksum);
+
+} // namespace file_cmds
+
+} // namespace platform_mc
+
+} // namespace pldm
